### PR TITLE
Detailed Health Check Update

### PR DESCRIPTION
**Pull request by ZairMahmood**
## Code reviewers

- [x] @edk 

## Summary of change

Added in detailed health check
 - Updated base to either require proc or no procs
 - Assign each class to inherit from BaseNoProc or BaseProc
 - Overwrite call functions on BaseNoProc
 - Create separation between simple health checks and detailed checks
 - Rename redis_check to resque_check
 - Add in scheduler check
 - removed usage of hard_fail (no purpose?)
 - improved some memoizations 
 - usage of awesome yield + super blocks

## Testing approach

Manually tested on related CP microservices

### CODE REVIEW

File: lib/simple_health_check/base_no_proc.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        connection = yield
+      end
+      @response_time = Time.now - start_time
+      status = connection ? :ok : :crit
+      response.status_code = status
+    rescue
+      # catch exceptions since we don't want the health-check to bubble all the way to the top
+      status = :crit
+      error = $ERROR_INFO.to_s
+      response.status_code = :crit
```
Comment (edk) [inline lib/simple_health_check/base_no_proc.rb]: i'm not aware of this status code, can we use one of the standard ones here? https://gist.github.com/mlanett/a31c340b132ddefa9cca

File: lib/simple_health_check/base_no_proc.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+class SimpleHealthCheck::BaseNoProc < SimpleHealthCheck::Base
+  def initialize service_name: nil, check_proc: nil
+    super
+  end
+
+  # proc-not required call
+  def call(response:)
+    status = nil
```
Comment (edk) [inline lib/simple_health_check/base_no_proc.rb]: usually an assignment like this isn't usually necessary as variable assignments are hoisted and set in a function scope.  As long as a variable is set somewhere, even if the code branch isn't executed, the default would be automatically nil.

https://ieftimov.com/post/variable-hoisting-ruby/ might help explain better than i can.

File: lib/simple_health_check/base_no_proc.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      @response_time = Time.now - start_time
+      status = connection ? :ok : :crit
+      response.status_code = status
+    rescue
+      # catch exceptions since we don't want the health-check to bubble all the way to the top
+      status = :crit
+      error = $ERROR_INFO.to_s
+      response.status_code = :crit
+    end
+    [status, error]
```
Comment (edk) [inline lib/simple_health_check/base_no_proc.rb]: doing a `[status, error.to_s]` might let us skip the error assignment at the top as well as the `.to_s` on the error objects elsewhere in this method.  This would ensure we never return a nil object, and always an empty string instead.

File: lib/simple_health_check/mysql_check.rb [hunk 4]
```diff
@@ -1,38 +1,23 @@
-class SimpleHealthCheck::MysqlCheck < SimpleHealthCheck::Base
-  def initialize service_name: "msyql", check_proc: nil, hard_fail: false
```
Comment (edk) [inline lib/simple_health_check/mysql_check.rb]: whoops :D

File: lib/simple_health_check/mysql_check.rb [hunk 5]
```diff
@@ -1,38 +1,23 @@
...
-    @service_name = service_name
+class SimpleHealthCheck::MysqlCheck < SimpleHealthCheck::BaseNoProc
+  def initialize service_name: "mysql", check_proc: nil
+    @service_name = service_name.to_s
```
Comment (edk) [inline lib/simple_health_check/mysql_check.rb]: curious, were you getting non-string values to this argument?

File: lib/simple_health_check/mysql_check.rb [hunk 6]
```diff
@@ -1,38 +1,23 @@
...
-    if @proc && @proc.respond_to?(:call)
+    if @version.nil?
       begin
-        # @proc is a user-supplied function to see if mysql connection is working.
-        # It can be as simple as `User.first`
-        # If it doesn't throw an exception, we're good
-        rv = @proc.call
-        response.add name: @service_name, status: rv
-        response.status_code = rv
+        @version = @version.nil? ? version_check : @version
```
Comment (edk) [inline lib/simple_health_check/mysql_check.rb]: since you've already checked for `version.nil?` in the if expression, this can probably be simplified to just `@version = version_check`

File: lib/simple_health_check/scheduler_check.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+  def initialize service_name: "scheduler", check_proc: nil, time_frame: nil
+    @service_name = service_name
+    @proc = check_proc || SimpleHealthCheck::Configuration.scheduler_check_proc
+    @type = 'internal'
+    @time_frame = time_frame
+  end
+
+  def call(response:)
+    super {
+      ((Time.now - JobTracker.last.updated_at) <= @time_frame.to_i) unless @time_frame.nil?
```
Comment (edk) [inline lib/simple_health_check/scheduler_check.rb]: not every project that includes this gem will have a `JobTracker` model.  Is this just an example, or will it cause issues for other projects that use this dependency?

File: lib/simple_health_check/version_check.rb [hunk 8]
```diff
@@ -3,15 +3,15 @@
...
 class SimpleHealthCheck::VersionCheck < SimpleHealthCheck::Base
   def call(response:)
+    # memoize to avoid repeatedly reading from static file
+    @ver = nil
     if @ver.nil?
-      # memoize to avoid repeatedly reading from static file
-      @ver = "unknown"
       if File.exist?(SimpleHealthCheck::Configuration.version_file)
-        @ver = File.read(SimpleHealthCheck::Configuration.version_file).strip rescue "unknown"
+        @ver = File.read(SimpleHealthCheck::Configuration.version_file).strip rescue ''
```
Comment (edk) [inline lib/simple_health_check/version_check.rb]: why this change?

Comment (ZairMahmood) [reply lib/simple_health_check/version_check.rb]: Personal preference, I can change this back

File: lib/simple_health_check/scheduler_check.rb (hunk 7 above)
Comment (ZairMahmood) [reply lib/simple_health_check/scheduler_check.rb]: It won't cause issues because this will be caught by a rescue block in projects that don't have JobTracker and the detailed health check will return crit for it. It can also be avoided in projects by not adding the SchedulerCheck configuration. The other option is also passing in your own proc method to call.

File: lib/simple_health_check/mysql_check.rb (hunk 6 above)
Comment (ZairMahmood) [reply lib/simple_health_check/mysql_check.rb]: 👍

File: lib/simple_health_check/mysql_check.rb (hunk 5 above)
Comment (ZairMahmood) [reply lib/simple_health_check/mysql_check.rb]: There was no issue here. I thought there was when I changed this but it was related to something else.

File: lib/simple_health_check/base_no_proc.rb (hunk 3 above)
Comment (ZairMahmood) [reply lib/simple_health_check/base_no_proc.rb]: 👍

File: lib/simple_health_check/base_no_proc.rb (hunk 2 above)
Comment (ZairMahmood) [reply lib/simple_health_check/base_no_proc.rb]: I can remove this. 👍

File: lib/simple_health_check/base_no_proc.rb (hunk 1 above)
Comment (ZairMahmood) [reply lib/simple_health_check/base_no_proc.rb]: This status_code was used to iterate thru all services such that the status didn't change back to OK from CRIT. I should probably rename this to overall_status and in my controller, return 200

File: lib/simple_health_check/mysql_check.rb (hunk 6 above)
Comment (mrpasquini) [reply lib/simple_health_check/mysql_check.rb]: Cant this be simplified even more to just be the short circuited assignment?
```
begin
  @version ||= version_check
rescue
  @version = nil
```
basically if `@version` evals to true we dont evaluate the other half of the `||`

File: lib/simple_health_check/scheduler_check.rb (hunk 7 above)
Comment (mrpasquini) [reply lib/simple_health_check/scheduler_check.rb]: Yeah, would this fail for eager loaded projects (ie production not development mode)? I'm not sure it would get caught.  Can the class of `JobTracker` be passed into this as a variable? 

Also `JobTracker.last` is going to be the `JobTracker` with the largest ID, not necesarily the biggest value for `updated_at`.

Comment (ZairMahmood) [reply lib/simple_health_check/scheduler_check.rb]: It won't cause the application to fail from launching. The status will return as crit in the detailed health check with an error most likely. This line of code is being rescued by the rescue in `base_no_proc.rb`

So for JobTracker, would it be better to find the last job run based on the created_at timestamp?

Comment (mrpasquini) [reply lib/simple_health_check/scheduler_check.rb]: Yeah i guess you would even catch loading errors on startup and continue... i'd be worried that would lead to weird things but yeah maybe its ok.



functionally idk whats best here to search on here. I probably would make this check more generic instead of tightly tied to coupa pay and JobTracker.  You could pass in the model_class and the field_name you want to check and its generically a checker to the newest record.
You can use `maximum` on an class or AREL object and it will build the query in sql to find the value.

```
model_class.send(:maximum, field_name)
```

which would query like
```
   (0.6ms)  SELECT MAX(`users`.`created_at`) FROM `users`
=> Fri, 08 Feb 2019 19:08:51 UTC +00:00
```
and return it as a `ActiveSupport::TimeWithZone` DateTime object.

Comment (ZairMahmood) [reply lib/simple_health_check/scheduler_check.rb]: how about this:
```
(byebug) JobTracker.send(:maximum, :updated_at)
Fri, 12 Apr 2019 21:48:27 UTC +00:00
(byebug) Time.now - JobTracker.send(:maximum, :updated_at)
336984.268918
(byebug) @time_frame
30 minutes
(byebug) @time_frame.to_i
1800
(byebug) Time.now - JobTracker.send(:maximum, :updated_at) <= @time_frame.to_i
false
```

pretty much the last line

File: lib/simple_health_check.rb [hunk 9]
```diff
@@ -1,9 +1,21 @@
...
+    http_endpoint_check
+    json_file
+    mysql_check
+    resque_check
+    scheduler_check
+    s3_check
+    version_check
+    version
+  ].each do |file|
+    classified_string = file.split('_').collect!(&:capitalize).join
```
Comment (edk) [inline lib/simple_health_check.rb]: fyi, since we assume this gem will be used in a rails app, there is also a rails (activesupport) method that does cliassify in one shot: https://apidock.com/rails/ActiveSupport/Inflector/classify

However, if this gem is ever used outside of a rails project, that won't be available.

In my mind, this gem can be restricted to just rails projects.

**Comment by edk** [comment]
👍
